File: DiagonalIteration.cpp
//Time complexity: O(m*n)
//Space complexity O(1)
//Approach: we are trying to maintain a flag either 1 or -1
// 1: bottom to top -1: top to bottom
//we will always start from mat[0][0] and resultant array would be sized M*N
//everytime we hit row 0 

class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
      
      int m=mat.size();
      int n=mat[0].size();
      int index=0;
      int dir=1;
      int r=0, c=0;
        vector<int> res(m*n);
        while(index<m*n)
        {
            res[index]=mat[r][c];
            index++;
            if(dir==1) //going bottom to top
            {
                if(c==n-1)
                {
                    dir=-1;
                    r++;
                }
                else if(r==0)
                {
                    dir=-1;
                    c++;
                }              
                else
                {
                    r--; c++;
                }
            }
            else //going top to bottom
            {
                if(r==m-1)
                {
                    dir=1;
                    c++;
                }
                else if(c==0)
                {
                    r++;
                    dir=1;
                }
                else
                {
                    r++; c--;
                }
            }
        }
            return res;
        

      
      
      
      /*  
        int n=mat.size();
        int m=mat[0].size();
        vector<int> ans;
        map<int, vector<int>> mp;
        
        for(int i=0;i<mat.size();i++)
        {
            for(int j=0;j<mat[0].size();j++)
            {
                mp[i+j].push_back(mat[i][j]);
            }
        }
        
        for(auto i: mp)
        {
            if(i.first%2==0)
                reverse(i.second.begin(), i.second.end());
        for(auto x: i.second)
            ans.push_back(x);
        }
        
        return ans;*/
    }
};

File: diagonalIteration.java
// Time Complexity = O(m*n)
// Space Complexity = O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Logic: Here we iterate over the input matrix and manipulate the indexes and filling the result array.
// We need to change course of the upward traversal when we hit one of these 2 cases: 1) Its the last coloumn 2) Its the 0th row
// We need to change course of the downward traversal when we hit one of these 2 cases: 1) Its the first coloumn 2) Its the last (n-1) row

// Your code here along with comments explaining your approach
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        if (mat == null || mat.length == 0) return new int[]{};
        int i=0, j=0;
        int dir=1;  // Initialize the direction to 1 since we will go upwards first
        int m=mat.length;
        int n=mat[0].length;
        int[] result = new int[m*n];

        // Itterate over the mat array and update the result array by manipulating i, j and index value
        for (int index=0; index < m*n; index++) {
            // Update the result with the input matrix at i, j index
            result[index] = mat[i][j];

            // For the upward direction
            if (dir == 1) {
                if (j == n-1) {
                    dir = -1;
                    i++;
                }
                else if (i == 0) {
                    dir = -1;
                    j++;
                }
                else {
                    j++;
                    i--;
                }
            }
            // For the downward direction
            else {
                if (i == m-1) {
                    dir = 1;
                    j++;
                }
                else if (j == 0) {
                    dir = 1;
                    i++;
                }
                else {
                    i++;
                    j--;
                }
            }
        }

        return result;
    }
}